// Code generated by software.amazon.smithy.rust.codegen.smithy-rs. DO NOT EDIT.
use crate::model::AccessDeniedExceptionReason;
use crate::model::RequestThrottledExceptionReason;
use crate::model::ResourceNotFoundExceptionReason;
use crate::model::ServiceQuotaExceededExceptionReason;
use crate::model::ValidationExceptionReason;
#[non_exhaustive]
#[derive(::std::fmt::Debug)]
pub enum CompleteSnapshotError {
    AccessDeniedError(AccessDeniedError),
    InternalServerError(InternalServerError),
    RequestThrottledError(RequestThrottledError),
    ResourceNotFoundError(ResourceNotFoundError),
    ServiceQuotaExceededError(ServiceQuotaExceededError),
    ValidationError(ValidationError),

    /// An unexpected error, eg. invalid JSON returned by the service
    Unhandled(Box<dyn ::std::error::Error>),
}
impl ::std::fmt::Display for CompleteSnapshotError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CompleteSnapshotError::AccessDeniedError(_inner) => _inner.fmt(f),
            CompleteSnapshotError::InternalServerError(_inner) => _inner.fmt(f),
            CompleteSnapshotError::RequestThrottledError(_inner) => _inner.fmt(f),
            CompleteSnapshotError::ResourceNotFoundError(_inner) => _inner.fmt(f),
            CompleteSnapshotError::ServiceQuotaExceededError(_inner) => _inner.fmt(f),
            CompleteSnapshotError::ValidationError(_inner) => _inner.fmt(f),
            CompleteSnapshotError::Unhandled(_inner) => _inner.fmt(f),
        }
    }
}
impl CompleteSnapshotError {
    pub fn unhandled<E: Into<Box<dyn ::std::error::Error>>>(err: E) -> Self {
        CompleteSnapshotError::Unhandled(err.into())
    }
    pub fn message(&self) -> Option<&str> {
        match self {
            CompleteSnapshotError::AccessDeniedError(_inner) => _inner.message(),
            CompleteSnapshotError::InternalServerError(_inner) => _inner.message(),
            CompleteSnapshotError::RequestThrottledError(_inner) => _inner.message(),
            CompleteSnapshotError::ResourceNotFoundError(_inner) => _inner.message(),
            CompleteSnapshotError::ServiceQuotaExceededError(_inner) => _inner.message(),
            CompleteSnapshotError::ValidationError(_inner) => _inner.message(),
            CompleteSnapshotError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.message(),
                    None => None,
                }
            }
        }
    }
    pub fn code(&self) -> Option<&str> {
        match self {
            CompleteSnapshotError::AccessDeniedError(_inner) => Some(_inner.code()),
            CompleteSnapshotError::InternalServerError(_inner) => Some(_inner.code()),
            CompleteSnapshotError::RequestThrottledError(_inner) => Some(_inner.code()),
            CompleteSnapshotError::ResourceNotFoundError(_inner) => Some(_inner.code()),
            CompleteSnapshotError::ServiceQuotaExceededError(_inner) => Some(_inner.code()),
            CompleteSnapshotError::ValidationError(_inner) => Some(_inner.code()),
            CompleteSnapshotError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.code(),
                    None => None,
                }
            }
        }
    }
}
impl ::std::error::Error for CompleteSnapshotError {
    fn source(&self) -> Option<&(dyn ::std::error::Error + 'static)> {
        match self {
            CompleteSnapshotError::AccessDeniedError(_inner) => Some(_inner),
            CompleteSnapshotError::InternalServerError(_inner) => Some(_inner),
            CompleteSnapshotError::RequestThrottledError(_inner) => Some(_inner),
            CompleteSnapshotError::ResourceNotFoundError(_inner) => Some(_inner),
            CompleteSnapshotError::ServiceQuotaExceededError(_inner) => Some(_inner),
            CompleteSnapshotError::ValidationError(_inner) => Some(_inner),
            CompleteSnapshotError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => Some(_inner),
                    None => Some(_inner.as_ref()),
                }
            }
        }
    }
}

#[non_exhaustive]
#[derive(::std::fmt::Debug)]
pub enum GetSnapshotBlockError {
    AccessDeniedError(AccessDeniedError),
    InternalServerError(InternalServerError),
    RequestThrottledError(RequestThrottledError),
    ResourceNotFoundError(ResourceNotFoundError),
    ServiceQuotaExceededError(ServiceQuotaExceededError),
    ValidationError(ValidationError),

    /// An unexpected error, eg. invalid JSON returned by the service
    Unhandled(Box<dyn ::std::error::Error>),
}
impl ::std::fmt::Display for GetSnapshotBlockError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            GetSnapshotBlockError::AccessDeniedError(_inner) => _inner.fmt(f),
            GetSnapshotBlockError::InternalServerError(_inner) => _inner.fmt(f),
            GetSnapshotBlockError::RequestThrottledError(_inner) => _inner.fmt(f),
            GetSnapshotBlockError::ResourceNotFoundError(_inner) => _inner.fmt(f),
            GetSnapshotBlockError::ServiceQuotaExceededError(_inner) => _inner.fmt(f),
            GetSnapshotBlockError::ValidationError(_inner) => _inner.fmt(f),
            GetSnapshotBlockError::Unhandled(_inner) => _inner.fmt(f),
        }
    }
}
impl GetSnapshotBlockError {
    pub fn unhandled<E: Into<Box<dyn ::std::error::Error>>>(err: E) -> Self {
        GetSnapshotBlockError::Unhandled(err.into())
    }
    pub fn message(&self) -> Option<&str> {
        match self {
            GetSnapshotBlockError::AccessDeniedError(_inner) => _inner.message(),
            GetSnapshotBlockError::InternalServerError(_inner) => _inner.message(),
            GetSnapshotBlockError::RequestThrottledError(_inner) => _inner.message(),
            GetSnapshotBlockError::ResourceNotFoundError(_inner) => _inner.message(),
            GetSnapshotBlockError::ServiceQuotaExceededError(_inner) => _inner.message(),
            GetSnapshotBlockError::ValidationError(_inner) => _inner.message(),
            GetSnapshotBlockError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.message(),
                    None => None,
                }
            }
        }
    }
    pub fn code(&self) -> Option<&str> {
        match self {
            GetSnapshotBlockError::AccessDeniedError(_inner) => Some(_inner.code()),
            GetSnapshotBlockError::InternalServerError(_inner) => Some(_inner.code()),
            GetSnapshotBlockError::RequestThrottledError(_inner) => Some(_inner.code()),
            GetSnapshotBlockError::ResourceNotFoundError(_inner) => Some(_inner.code()),
            GetSnapshotBlockError::ServiceQuotaExceededError(_inner) => Some(_inner.code()),
            GetSnapshotBlockError::ValidationError(_inner) => Some(_inner.code()),
            GetSnapshotBlockError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.code(),
                    None => None,
                }
            }
        }
    }
}
impl ::std::error::Error for GetSnapshotBlockError {
    fn source(&self) -> Option<&(dyn ::std::error::Error + 'static)> {
        match self {
            GetSnapshotBlockError::AccessDeniedError(_inner) => Some(_inner),
            GetSnapshotBlockError::InternalServerError(_inner) => Some(_inner),
            GetSnapshotBlockError::RequestThrottledError(_inner) => Some(_inner),
            GetSnapshotBlockError::ResourceNotFoundError(_inner) => Some(_inner),
            GetSnapshotBlockError::ServiceQuotaExceededError(_inner) => Some(_inner),
            GetSnapshotBlockError::ValidationError(_inner) => Some(_inner),
            GetSnapshotBlockError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => Some(_inner),
                    None => Some(_inner.as_ref()),
                }
            }
        }
    }
}

#[non_exhaustive]
#[derive(::std::fmt::Debug)]
pub enum ListChangedBlocksError {
    AccessDeniedError(AccessDeniedError),
    InternalServerError(InternalServerError),
    RequestThrottledError(RequestThrottledError),
    ResourceNotFoundError(ResourceNotFoundError),
    ServiceQuotaExceededError(ServiceQuotaExceededError),
    ValidationError(ValidationError),

    /// An unexpected error, eg. invalid JSON returned by the service
    Unhandled(Box<dyn ::std::error::Error>),
}
impl ::std::fmt::Display for ListChangedBlocksError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ListChangedBlocksError::AccessDeniedError(_inner) => _inner.fmt(f),
            ListChangedBlocksError::InternalServerError(_inner) => _inner.fmt(f),
            ListChangedBlocksError::RequestThrottledError(_inner) => _inner.fmt(f),
            ListChangedBlocksError::ResourceNotFoundError(_inner) => _inner.fmt(f),
            ListChangedBlocksError::ServiceQuotaExceededError(_inner) => _inner.fmt(f),
            ListChangedBlocksError::ValidationError(_inner) => _inner.fmt(f),
            ListChangedBlocksError::Unhandled(_inner) => _inner.fmt(f),
        }
    }
}
impl ListChangedBlocksError {
    pub fn unhandled<E: Into<Box<dyn ::std::error::Error>>>(err: E) -> Self {
        ListChangedBlocksError::Unhandled(err.into())
    }
    pub fn message(&self) -> Option<&str> {
        match self {
            ListChangedBlocksError::AccessDeniedError(_inner) => _inner.message(),
            ListChangedBlocksError::InternalServerError(_inner) => _inner.message(),
            ListChangedBlocksError::RequestThrottledError(_inner) => _inner.message(),
            ListChangedBlocksError::ResourceNotFoundError(_inner) => _inner.message(),
            ListChangedBlocksError::ServiceQuotaExceededError(_inner) => _inner.message(),
            ListChangedBlocksError::ValidationError(_inner) => _inner.message(),
            ListChangedBlocksError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.message(),
                    None => None,
                }
            }
        }
    }
    pub fn code(&self) -> Option<&str> {
        match self {
            ListChangedBlocksError::AccessDeniedError(_inner) => Some(_inner.code()),
            ListChangedBlocksError::InternalServerError(_inner) => Some(_inner.code()),
            ListChangedBlocksError::RequestThrottledError(_inner) => Some(_inner.code()),
            ListChangedBlocksError::ResourceNotFoundError(_inner) => Some(_inner.code()),
            ListChangedBlocksError::ServiceQuotaExceededError(_inner) => Some(_inner.code()),
            ListChangedBlocksError::ValidationError(_inner) => Some(_inner.code()),
            ListChangedBlocksError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.code(),
                    None => None,
                }
            }
        }
    }
}
impl ::std::error::Error for ListChangedBlocksError {
    fn source(&self) -> Option<&(dyn ::std::error::Error + 'static)> {
        match self {
            ListChangedBlocksError::AccessDeniedError(_inner) => Some(_inner),
            ListChangedBlocksError::InternalServerError(_inner) => Some(_inner),
            ListChangedBlocksError::RequestThrottledError(_inner) => Some(_inner),
            ListChangedBlocksError::ResourceNotFoundError(_inner) => Some(_inner),
            ListChangedBlocksError::ServiceQuotaExceededError(_inner) => Some(_inner),
            ListChangedBlocksError::ValidationError(_inner) => Some(_inner),
            ListChangedBlocksError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => Some(_inner),
                    None => Some(_inner.as_ref()),
                }
            }
        }
    }
}

#[non_exhaustive]
#[derive(::std::fmt::Debug)]
pub enum ListSnapshotBlocksError {
    AccessDeniedError(AccessDeniedError),
    InternalServerError(InternalServerError),
    RequestThrottledError(RequestThrottledError),
    ResourceNotFoundError(ResourceNotFoundError),
    ServiceQuotaExceededError(ServiceQuotaExceededError),
    ValidationError(ValidationError),

    /// An unexpected error, eg. invalid JSON returned by the service
    Unhandled(Box<dyn ::std::error::Error>),
}
impl ::std::fmt::Display for ListSnapshotBlocksError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ListSnapshotBlocksError::AccessDeniedError(_inner) => _inner.fmt(f),
            ListSnapshotBlocksError::InternalServerError(_inner) => _inner.fmt(f),
            ListSnapshotBlocksError::RequestThrottledError(_inner) => _inner.fmt(f),
            ListSnapshotBlocksError::ResourceNotFoundError(_inner) => _inner.fmt(f),
            ListSnapshotBlocksError::ServiceQuotaExceededError(_inner) => _inner.fmt(f),
            ListSnapshotBlocksError::ValidationError(_inner) => _inner.fmt(f),
            ListSnapshotBlocksError::Unhandled(_inner) => _inner.fmt(f),
        }
    }
}
impl ListSnapshotBlocksError {
    pub fn unhandled<E: Into<Box<dyn ::std::error::Error>>>(err: E) -> Self {
        ListSnapshotBlocksError::Unhandled(err.into())
    }
    pub fn message(&self) -> Option<&str> {
        match self {
            ListSnapshotBlocksError::AccessDeniedError(_inner) => _inner.message(),
            ListSnapshotBlocksError::InternalServerError(_inner) => _inner.message(),
            ListSnapshotBlocksError::RequestThrottledError(_inner) => _inner.message(),
            ListSnapshotBlocksError::ResourceNotFoundError(_inner) => _inner.message(),
            ListSnapshotBlocksError::ServiceQuotaExceededError(_inner) => _inner.message(),
            ListSnapshotBlocksError::ValidationError(_inner) => _inner.message(),
            ListSnapshotBlocksError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.message(),
                    None => None,
                }
            }
        }
    }
    pub fn code(&self) -> Option<&str> {
        match self {
            ListSnapshotBlocksError::AccessDeniedError(_inner) => Some(_inner.code()),
            ListSnapshotBlocksError::InternalServerError(_inner) => Some(_inner.code()),
            ListSnapshotBlocksError::RequestThrottledError(_inner) => Some(_inner.code()),
            ListSnapshotBlocksError::ResourceNotFoundError(_inner) => Some(_inner.code()),
            ListSnapshotBlocksError::ServiceQuotaExceededError(_inner) => Some(_inner.code()),
            ListSnapshotBlocksError::ValidationError(_inner) => Some(_inner.code()),
            ListSnapshotBlocksError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.code(),
                    None => None,
                }
            }
        }
    }
}
impl ::std::error::Error for ListSnapshotBlocksError {
    fn source(&self) -> Option<&(dyn ::std::error::Error + 'static)> {
        match self {
            ListSnapshotBlocksError::AccessDeniedError(_inner) => Some(_inner),
            ListSnapshotBlocksError::InternalServerError(_inner) => Some(_inner),
            ListSnapshotBlocksError::RequestThrottledError(_inner) => Some(_inner),
            ListSnapshotBlocksError::ResourceNotFoundError(_inner) => Some(_inner),
            ListSnapshotBlocksError::ServiceQuotaExceededError(_inner) => Some(_inner),
            ListSnapshotBlocksError::ValidationError(_inner) => Some(_inner),
            ListSnapshotBlocksError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => Some(_inner),
                    None => Some(_inner.as_ref()),
                }
            }
        }
    }
}

#[non_exhaustive]
#[derive(::std::fmt::Debug)]
pub enum PutSnapshotBlockError {
    AccessDeniedError(AccessDeniedError),
    InternalServerError(InternalServerError),
    RequestThrottledError(RequestThrottledError),
    ResourceNotFoundError(ResourceNotFoundError),
    ServiceQuotaExceededError(ServiceQuotaExceededError),
    ValidationError(ValidationError),

    /// An unexpected error, eg. invalid JSON returned by the service
    Unhandled(Box<dyn ::std::error::Error>),
}
impl ::std::fmt::Display for PutSnapshotBlockError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PutSnapshotBlockError::AccessDeniedError(_inner) => _inner.fmt(f),
            PutSnapshotBlockError::InternalServerError(_inner) => _inner.fmt(f),
            PutSnapshotBlockError::RequestThrottledError(_inner) => _inner.fmt(f),
            PutSnapshotBlockError::ResourceNotFoundError(_inner) => _inner.fmt(f),
            PutSnapshotBlockError::ServiceQuotaExceededError(_inner) => _inner.fmt(f),
            PutSnapshotBlockError::ValidationError(_inner) => _inner.fmt(f),
            PutSnapshotBlockError::Unhandled(_inner) => _inner.fmt(f),
        }
    }
}
impl PutSnapshotBlockError {
    pub fn unhandled<E: Into<Box<dyn ::std::error::Error>>>(err: E) -> Self {
        PutSnapshotBlockError::Unhandled(err.into())
    }
    pub fn message(&self) -> Option<&str> {
        match self {
            PutSnapshotBlockError::AccessDeniedError(_inner) => _inner.message(),
            PutSnapshotBlockError::InternalServerError(_inner) => _inner.message(),
            PutSnapshotBlockError::RequestThrottledError(_inner) => _inner.message(),
            PutSnapshotBlockError::ResourceNotFoundError(_inner) => _inner.message(),
            PutSnapshotBlockError::ServiceQuotaExceededError(_inner) => _inner.message(),
            PutSnapshotBlockError::ValidationError(_inner) => _inner.message(),
            PutSnapshotBlockError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.message(),
                    None => None,
                }
            }
        }
    }
    pub fn code(&self) -> Option<&str> {
        match self {
            PutSnapshotBlockError::AccessDeniedError(_inner) => Some(_inner.code()),
            PutSnapshotBlockError::InternalServerError(_inner) => Some(_inner.code()),
            PutSnapshotBlockError::RequestThrottledError(_inner) => Some(_inner.code()),
            PutSnapshotBlockError::ResourceNotFoundError(_inner) => Some(_inner.code()),
            PutSnapshotBlockError::ServiceQuotaExceededError(_inner) => Some(_inner.code()),
            PutSnapshotBlockError::ValidationError(_inner) => Some(_inner.code()),
            PutSnapshotBlockError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.code(),
                    None => None,
                }
            }
        }
    }
}
impl ::std::error::Error for PutSnapshotBlockError {
    fn source(&self) -> Option<&(dyn ::std::error::Error + 'static)> {
        match self {
            PutSnapshotBlockError::AccessDeniedError(_inner) => Some(_inner),
            PutSnapshotBlockError::InternalServerError(_inner) => Some(_inner),
            PutSnapshotBlockError::RequestThrottledError(_inner) => Some(_inner),
            PutSnapshotBlockError::ResourceNotFoundError(_inner) => Some(_inner),
            PutSnapshotBlockError::ServiceQuotaExceededError(_inner) => Some(_inner),
            PutSnapshotBlockError::ValidationError(_inner) => Some(_inner),
            PutSnapshotBlockError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => Some(_inner),
                    None => Some(_inner.as_ref()),
                }
            }
        }
    }
}

#[non_exhaustive]
#[derive(::std::fmt::Debug)]
pub enum StartSnapshotError {
    AccessDeniedError(AccessDeniedError),
    ConcurrentLimitExceededError(ConcurrentLimitExceededError),
    ConflictError(ConflictError),
    InternalServerError(InternalServerError),
    RequestThrottledError(RequestThrottledError),
    ResourceNotFoundError(ResourceNotFoundError),
    ServiceQuotaExceededError(ServiceQuotaExceededError),
    ValidationError(ValidationError),

    /// An unexpected error, eg. invalid JSON returned by the service
    Unhandled(Box<dyn ::std::error::Error>),
}
impl ::std::fmt::Display for StartSnapshotError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            StartSnapshotError::AccessDeniedError(_inner) => _inner.fmt(f),
            StartSnapshotError::ConcurrentLimitExceededError(_inner) => _inner.fmt(f),
            StartSnapshotError::ConflictError(_inner) => _inner.fmt(f),
            StartSnapshotError::InternalServerError(_inner) => _inner.fmt(f),
            StartSnapshotError::RequestThrottledError(_inner) => _inner.fmt(f),
            StartSnapshotError::ResourceNotFoundError(_inner) => _inner.fmt(f),
            StartSnapshotError::ServiceQuotaExceededError(_inner) => _inner.fmt(f),
            StartSnapshotError::ValidationError(_inner) => _inner.fmt(f),
            StartSnapshotError::Unhandled(_inner) => _inner.fmt(f),
        }
    }
}
impl StartSnapshotError {
    pub fn unhandled<E: Into<Box<dyn ::std::error::Error>>>(err: E) -> Self {
        StartSnapshotError::Unhandled(err.into())
    }
    pub fn message(&self) -> Option<&str> {
        match self {
            StartSnapshotError::AccessDeniedError(_inner) => _inner.message(),
            StartSnapshotError::ConcurrentLimitExceededError(_inner) => _inner.message(),
            StartSnapshotError::ConflictError(_inner) => _inner.message(),
            StartSnapshotError::InternalServerError(_inner) => _inner.message(),
            StartSnapshotError::RequestThrottledError(_inner) => _inner.message(),
            StartSnapshotError::ResourceNotFoundError(_inner) => _inner.message(),
            StartSnapshotError::ServiceQuotaExceededError(_inner) => _inner.message(),
            StartSnapshotError::ValidationError(_inner) => _inner.message(),
            StartSnapshotError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.message(),
                    None => None,
                }
            }
        }
    }
    pub fn code(&self) -> Option<&str> {
        match self {
            StartSnapshotError::AccessDeniedError(_inner) => Some(_inner.code()),
            StartSnapshotError::ConcurrentLimitExceededError(_inner) => Some(_inner.code()),
            StartSnapshotError::ConflictError(_inner) => Some(_inner.code()),
            StartSnapshotError::InternalServerError(_inner) => Some(_inner.code()),
            StartSnapshotError::RequestThrottledError(_inner) => Some(_inner.code()),
            StartSnapshotError::ResourceNotFoundError(_inner) => Some(_inner.code()),
            StartSnapshotError::ServiceQuotaExceededError(_inner) => Some(_inner.code()),
            StartSnapshotError::ValidationError(_inner) => Some(_inner.code()),
            StartSnapshotError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => _inner.code(),
                    None => None,
                }
            }
        }
    }
}
impl ::std::error::Error for StartSnapshotError {
    fn source(&self) -> Option<&(dyn ::std::error::Error + 'static)> {
        match self {
            StartSnapshotError::AccessDeniedError(_inner) => Some(_inner),
            StartSnapshotError::ConcurrentLimitExceededError(_inner) => Some(_inner),
            StartSnapshotError::ConflictError(_inner) => Some(_inner),
            StartSnapshotError::InternalServerError(_inner) => Some(_inner),
            StartSnapshotError::RequestThrottledError(_inner) => Some(_inner),
            StartSnapshotError::ResourceNotFoundError(_inner) => Some(_inner),
            StartSnapshotError::ServiceQuotaExceededError(_inner) => Some(_inner),
            StartSnapshotError::ValidationError(_inner) => Some(_inner),
            StartSnapshotError::Unhandled(_inner) => {
                match _inner.downcast_ref::<::smithy_types::Error>() {
                    Some(_inner) => Some(_inner),
                    None => Some(_inner.as_ref()),
                }
            }
        }
    }
}

/// <p>The input fails to satisfy the constraints of the EBS direct APIs.</p>
#[non_exhaustive]
#[derive(::std::clone::Clone, ::std::cmp::PartialEq, ::std::fmt::Debug)]
pub struct ValidationError {
    pub message: ::std::option::Option<::std::string::String>,
    /// <p>The reason for the validation exception.</p>
    pub reason: ::std::option::Option<ValidationExceptionReason>,
}
impl ValidationError {
    pub fn retryable(&self) -> bool {
        false
    }
    pub fn throttling(&self) -> bool {
        false
    }
    pub fn code(&self) -> &str {
        "ValidationException"
    }
    pub fn message(&self) -> Option<&str> {
        None
    }
}
impl ::std::fmt::Display for ValidationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "ValidationError")?;
        Ok(())
    }
}
impl ::std::error::Error for ValidationError {}
/// See [`ValidationError`](crate::error::ValidationError)
pub mod validation_error {

    use crate::error::ValidationError;
    use crate::model::ValidationExceptionReason;
    /// A builder for [`ValidationError`](crate::error::ValidationError)
    #[non_exhaustive]
    #[derive(Debug, Clone, Default)]
    pub struct Builder {
        message: ::std::option::Option<::std::string::String>,
        reason: ::std::option::Option<ValidationExceptionReason>,
    }
    impl Builder {
        pub fn message(mut self, inp: impl Into<::std::string::String>) -> Self {
            self.message = Some(inp.into());
            self
        }
        /// <p>The reason for the validation exception.</p>
        pub fn reason(mut self, inp: ValidationExceptionReason) -> Self {
            self.reason = Some(inp);
            self
        }
        /// Consumes the builder and constructs a [`ValidationError`](crate::error::ValidationError)
        pub fn build(self) -> ValidationError {
            ValidationError {
                message: self.message,
                reason: self.reason,
            }
        }
    }
}
impl ValidationError {
    /// Creates a new builder-style object to manufacture [`ValidationError`](crate::error::ValidationError)
    pub fn builder() -> crate::error::validation_error::Builder {
        crate::error::validation_error::Builder::default()
    }
}

/// <p>Your current service quotas do not allow you to perform this action.</p>
#[non_exhaustive]
#[derive(::std::clone::Clone, ::std::cmp::PartialEq, ::std::fmt::Debug)]
pub struct ServiceQuotaExceededError {
    /// <p>The reason for the exception.</p>
    pub reason: ::std::option::Option<ServiceQuotaExceededExceptionReason>,
    pub message: ::std::option::Option<::std::string::String>,
}
impl ServiceQuotaExceededError {
    pub fn retryable(&self) -> bool {
        false
    }
    pub fn throttling(&self) -> bool {
        false
    }
    pub fn code(&self) -> &str {
        "ServiceQuotaExceededException"
    }
    pub fn message(&self) -> Option<&str> {
        None
    }
}
impl ::std::fmt::Display for ServiceQuotaExceededError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "ServiceQuotaExceededError")?;
        Ok(())
    }
}
impl ::std::error::Error for ServiceQuotaExceededError {}
/// See [`ServiceQuotaExceededError`](crate::error::ServiceQuotaExceededError)
pub mod service_quota_exceeded_error {

    use crate::error::ServiceQuotaExceededError;
    use crate::model::ServiceQuotaExceededExceptionReason;
    /// A builder for [`ServiceQuotaExceededError`](crate::error::ServiceQuotaExceededError)
    #[non_exhaustive]
    #[derive(Debug, Clone, Default)]
    pub struct Builder {
        reason: ::std::option::Option<ServiceQuotaExceededExceptionReason>,
        message: ::std::option::Option<::std::string::String>,
    }
    impl Builder {
        /// <p>The reason for the exception.</p>
        pub fn reason(mut self, inp: ServiceQuotaExceededExceptionReason) -> Self {
            self.reason = Some(inp);
            self
        }
        pub fn message(mut self, inp: impl Into<::std::string::String>) -> Self {
            self.message = Some(inp.into());
            self
        }
        /// Consumes the builder and constructs a [`ServiceQuotaExceededError`](crate::error::ServiceQuotaExceededError)
        pub fn build(self) -> ServiceQuotaExceededError {
            ServiceQuotaExceededError {
                reason: self.reason,
                message: self.message,
            }
        }
    }
}
impl ServiceQuotaExceededError {
    /// Creates a new builder-style object to manufacture [`ServiceQuotaExceededError`](crate::error::ServiceQuotaExceededError)
    pub fn builder() -> crate::error::service_quota_exceeded_error::Builder {
        crate::error::service_quota_exceeded_error::Builder::default()
    }
}

/// <p>The specified resource does not exist.</p>
#[non_exhaustive]
#[derive(::std::clone::Clone, ::std::cmp::PartialEq, ::std::fmt::Debug)]
pub struct ResourceNotFoundError {
    pub message: ::std::option::Option<::std::string::String>,
    /// <p>The reason for the exception.</p>
    pub reason: ::std::option::Option<ResourceNotFoundExceptionReason>,
}
impl ResourceNotFoundError {
    pub fn retryable(&self) -> bool {
        false
    }
    pub fn throttling(&self) -> bool {
        false
    }
    pub fn code(&self) -> &str {
        "ResourceNotFoundException"
    }
    pub fn message(&self) -> Option<&str> {
        None
    }
}
impl ::std::fmt::Display for ResourceNotFoundError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "ResourceNotFoundError")?;
        Ok(())
    }
}
impl ::std::error::Error for ResourceNotFoundError {}
/// See [`ResourceNotFoundError`](crate::error::ResourceNotFoundError)
pub mod resource_not_found_error {

    use crate::error::ResourceNotFoundError;
    use crate::model::ResourceNotFoundExceptionReason;
    /// A builder for [`ResourceNotFoundError`](crate::error::ResourceNotFoundError)
    #[non_exhaustive]
    #[derive(Debug, Clone, Default)]
    pub struct Builder {
        message: ::std::option::Option<::std::string::String>,
        reason: ::std::option::Option<ResourceNotFoundExceptionReason>,
    }
    impl Builder {
        pub fn message(mut self, inp: impl Into<::std::string::String>) -> Self {
            self.message = Some(inp.into());
            self
        }
        /// <p>The reason for the exception.</p>
        pub fn reason(mut self, inp: ResourceNotFoundExceptionReason) -> Self {
            self.reason = Some(inp);
            self
        }
        /// Consumes the builder and constructs a [`ResourceNotFoundError`](crate::error::ResourceNotFoundError)
        pub fn build(self) -> ResourceNotFoundError {
            ResourceNotFoundError {
                message: self.message,
                reason: self.reason,
            }
        }
    }
}
impl ResourceNotFoundError {
    /// Creates a new builder-style object to manufacture [`ResourceNotFoundError`](crate::error::ResourceNotFoundError)
    pub fn builder() -> crate::error::resource_not_found_error::Builder {
        crate::error::resource_not_found_error::Builder::default()
    }
}

/// <p>The number of API requests has exceed the maximum allowed API request throttling
/// limit.</p>
#[non_exhaustive]
#[derive(::std::clone::Clone, ::std::cmp::PartialEq, ::std::fmt::Debug)]
pub struct RequestThrottledError {
    pub message: ::std::option::Option<::std::string::String>,
    /// <p>The reason for the exception.</p>
    pub reason: ::std::option::Option<RequestThrottledExceptionReason>,
}
impl RequestThrottledError {
    pub fn retryable(&self) -> bool {
        false
    }
    pub fn throttling(&self) -> bool {
        false
    }
    pub fn code(&self) -> &str {
        "RequestThrottledException"
    }
    pub fn message(&self) -> Option<&str> {
        None
    }
}
impl ::std::fmt::Display for RequestThrottledError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "RequestThrottledError")?;
        Ok(())
    }
}
impl ::std::error::Error for RequestThrottledError {}
/// See [`RequestThrottledError`](crate::error::RequestThrottledError)
pub mod request_throttled_error {

    use crate::error::RequestThrottledError;
    use crate::model::RequestThrottledExceptionReason;
    /// A builder for [`RequestThrottledError`](crate::error::RequestThrottledError)
    #[non_exhaustive]
    #[derive(Debug, Clone, Default)]
    pub struct Builder {
        message: ::std::option::Option<::std::string::String>,
        reason: ::std::option::Option<RequestThrottledExceptionReason>,
    }
    impl Builder {
        pub fn message(mut self, inp: impl Into<::std::string::String>) -> Self {
            self.message = Some(inp.into());
            self
        }
        /// <p>The reason for the exception.</p>
        pub fn reason(mut self, inp: RequestThrottledExceptionReason) -> Self {
            self.reason = Some(inp);
            self
        }
        /// Consumes the builder and constructs a [`RequestThrottledError`](crate::error::RequestThrottledError)
        pub fn build(self) -> RequestThrottledError {
            RequestThrottledError {
                message: self.message,
                reason: self.reason,
            }
        }
    }
}
impl RequestThrottledError {
    /// Creates a new builder-style object to manufacture [`RequestThrottledError`](crate::error::RequestThrottledError)
    pub fn builder() -> crate::error::request_throttled_error::Builder {
        crate::error::request_throttled_error::Builder::default()
    }
}

/// <p>An internal error has occurred.</p>
#[non_exhaustive]
#[derive(::std::clone::Clone, ::std::cmp::PartialEq, ::std::fmt::Debug)]
pub struct InternalServerError {
    pub message: ::std::option::Option<::std::string::String>,
}
impl InternalServerError {
    pub fn retryable(&self) -> bool {
        false
    }
    pub fn throttling(&self) -> bool {
        false
    }
    pub fn code(&self) -> &str {
        "InternalServerException"
    }
    pub fn message(&self) -> Option<&str> {
        None
    }
}
impl ::std::fmt::Display for InternalServerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "InternalServerError")?;
        Ok(())
    }
}
impl ::std::error::Error for InternalServerError {}
/// See [`InternalServerError`](crate::error::InternalServerError)
pub mod internal_server_error {

    use crate::error::InternalServerError;
    /// A builder for [`InternalServerError`](crate::error::InternalServerError)
    #[non_exhaustive]
    #[derive(Debug, Clone, Default)]
    pub struct Builder {
        message: ::std::option::Option<::std::string::String>,
    }
    impl Builder {
        pub fn message(mut self, inp: impl Into<::std::string::String>) -> Self {
            self.message = Some(inp.into());
            self
        }
        /// Consumes the builder and constructs a [`InternalServerError`](crate::error::InternalServerError)
        pub fn build(self) -> InternalServerError {
            InternalServerError {
                message: self.message,
            }
        }
    }
}
impl InternalServerError {
    /// Creates a new builder-style object to manufacture [`InternalServerError`](crate::error::InternalServerError)
    pub fn builder() -> crate::error::internal_server_error::Builder {
        crate::error::internal_server_error::Builder::default()
    }
}

/// <p>The request uses the same client token as a previous, but non-identical
/// request.</p>
#[non_exhaustive]
#[derive(::std::clone::Clone, ::std::cmp::PartialEq, ::std::fmt::Debug)]
pub struct ConflictError {
    pub message: ::std::option::Option<::std::string::String>,
}
impl ConflictError {
    pub fn retryable(&self) -> bool {
        false
    }
    pub fn throttling(&self) -> bool {
        false
    }
    pub fn code(&self) -> &str {
        "ConflictException"
    }
    pub fn message(&self) -> Option<&str> {
        None
    }
}
impl ::std::fmt::Display for ConflictError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "ConflictError")?;
        Ok(())
    }
}
impl ::std::error::Error for ConflictError {}
/// See [`ConflictError`](crate::error::ConflictError)
pub mod conflict_error {

    use crate::error::ConflictError;
    /// A builder for [`ConflictError`](crate::error::ConflictError)
    #[non_exhaustive]
    #[derive(Debug, Clone, Default)]
    pub struct Builder {
        message: ::std::option::Option<::std::string::String>,
    }
    impl Builder {
        pub fn message(mut self, inp: impl Into<::std::string::String>) -> Self {
            self.message = Some(inp.into());
            self
        }
        /// Consumes the builder and constructs a [`ConflictError`](crate::error::ConflictError)
        pub fn build(self) -> ConflictError {
            ConflictError {
                message: self.message,
            }
        }
    }
}
impl ConflictError {
    /// Creates a new builder-style object to manufacture [`ConflictError`](crate::error::ConflictError)
    pub fn builder() -> crate::error::conflict_error::Builder {
        crate::error::conflict_error::Builder::default()
    }
}

/// <p>You have reached the limit for concurrent API requests. For more information, see
/// <a href="https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-accessing-snapshot.html#ebsapi-performance">Optimizing performance of the EBS direct APIs</a> in the <i>Amazon Elastic Compute Cloud
/// User Guide</i>.</p>
#[non_exhaustive]
#[derive(::std::clone::Clone, ::std::cmp::PartialEq, ::std::fmt::Debug)]
pub struct ConcurrentLimitExceededError {
    pub message: ::std::option::Option<::std::string::String>,
}
impl ConcurrentLimitExceededError {
    pub fn retryable(&self) -> bool {
        false
    }
    pub fn throttling(&self) -> bool {
        false
    }
    pub fn code(&self) -> &str {
        "ConcurrentLimitExceededException"
    }
    pub fn message(&self) -> Option<&str> {
        None
    }
}
impl ::std::fmt::Display for ConcurrentLimitExceededError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "ConcurrentLimitExceededError")?;
        Ok(())
    }
}
impl ::std::error::Error for ConcurrentLimitExceededError {}
/// See [`ConcurrentLimitExceededError`](crate::error::ConcurrentLimitExceededError)
pub mod concurrent_limit_exceeded_error {

    use crate::error::ConcurrentLimitExceededError;
    /// A builder for [`ConcurrentLimitExceededError`](crate::error::ConcurrentLimitExceededError)
    #[non_exhaustive]
    #[derive(Debug, Clone, Default)]
    pub struct Builder {
        message: ::std::option::Option<::std::string::String>,
    }
    impl Builder {
        pub fn message(mut self, inp: impl Into<::std::string::String>) -> Self {
            self.message = Some(inp.into());
            self
        }
        /// Consumes the builder and constructs a [`ConcurrentLimitExceededError`](crate::error::ConcurrentLimitExceededError)
        pub fn build(self) -> ConcurrentLimitExceededError {
            ConcurrentLimitExceededError {
                message: self.message,
            }
        }
    }
}
impl ConcurrentLimitExceededError {
    /// Creates a new builder-style object to manufacture [`ConcurrentLimitExceededError`](crate::error::ConcurrentLimitExceededError)
    pub fn builder() -> crate::error::concurrent_limit_exceeded_error::Builder {
        crate::error::concurrent_limit_exceeded_error::Builder::default()
    }
}

/// <p>You do not have sufficient access to perform this action.</p>
#[non_exhaustive]
#[derive(::std::clone::Clone, ::std::cmp::PartialEq, ::std::fmt::Debug)]
pub struct AccessDeniedError {
    pub message: ::std::option::Option<::std::string::String>,
    /// <p>The reason for the exception.</p>
    pub reason: ::std::option::Option<AccessDeniedExceptionReason>,
}
impl AccessDeniedError {
    pub fn retryable(&self) -> bool {
        false
    }
    pub fn throttling(&self) -> bool {
        false
    }
    pub fn code(&self) -> &str {
        "AccessDeniedException"
    }
    pub fn message(&self) -> Option<&str> {
        None
    }
}
impl ::std::fmt::Display for AccessDeniedError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "AccessDeniedError")?;
        Ok(())
    }
}
impl ::std::error::Error for AccessDeniedError {}
/// See [`AccessDeniedError`](crate::error::AccessDeniedError)
pub mod access_denied_error {

    use crate::error::AccessDeniedError;
    use crate::model::AccessDeniedExceptionReason;
    /// A builder for [`AccessDeniedError`](crate::error::AccessDeniedError)
    #[non_exhaustive]
    #[derive(Debug, Clone, Default)]
    pub struct Builder {
        message: ::std::option::Option<::std::string::String>,
        reason: ::std::option::Option<AccessDeniedExceptionReason>,
    }
    impl Builder {
        pub fn message(mut self, inp: impl Into<::std::string::String>) -> Self {
            self.message = Some(inp.into());
            self
        }
        /// <p>The reason for the exception.</p>
        pub fn reason(mut self, inp: AccessDeniedExceptionReason) -> Self {
            self.reason = Some(inp);
            self
        }
        /// Consumes the builder and constructs a [`AccessDeniedError`](crate::error::AccessDeniedError)
        pub fn build(self) -> AccessDeniedError {
            AccessDeniedError {
                message: self.message,
                reason: self.reason,
            }
        }
    }
}
impl AccessDeniedError {
    /// Creates a new builder-style object to manufacture [`AccessDeniedError`](crate::error::AccessDeniedError)
    pub fn builder() -> crate::error::access_denied_error::Builder {
        crate::error::access_denied_error::Builder::default()
    }
}
